Platform/Authentication/Token Provider/DefaultTokenProvider.swift

```diff
@@ -0,0 +1,228 @@
+            throw AuthenticationError.failedToSerializeBody
+        }
+        
+        request.httpBody = httpBody
+        
+        return request
+    }
+    
+    private func requestURL(with baseURL: URL, queryItems: [URLQueryItem]?) throws -> URL {
+        guard var components = URLComponents(url: self.url, resolvingAgainstBaseURL: false) else {
+            self.logger?.log("\(String(describing: self)) failed to build request URL.", logLevel: .error)
+            throw AuthenticationError.invalidURL
+        }
+        
+        if let queryItems = self.queryItems {
+            if components.queryItems != nil {
+                components.queryItems?.append(contentsOf: queryItems)
+            }
+            else {
+                components.queryItems = queryItems
+            }
+        }
+        
+        guard let requestURL = components.url else {
+            self.logger?.log("\(String(describing: self)) failed to build request URL.", logLevel: .error)
+            throw AuthenticationError.invalidURL
+        }
+        
+        return requestURL
+    }
+    
+    private func requestHeaders(for headers: [String : String]?) -> [String : String] {
+        var requestHeaders = [Header.Field.contentType.rawValue : Header.Value.applicationFormURLEncoded.rawValue]
+        
+        headers?.forEach { field, value in
+            requestHeaders[field] = value
+        }
+        
+        return requestHeaders
+    }
+    
+    private func requestBody(for body: [URLEncodedBodyItem]?) -> [URLEncodedBodyItem] {
+        let grantType = URLEncodedBodyItem(name: URLEncodedBodyItem.Name.grantType, value: URLEncodedBodyItem.Value.clientCredentials)
+        var requestBody = [grantType]
+        
+        if let body = body {
+            let containsGrantType = body.contains { item -> Bool in
+                return item.name == URLEncodedBodyItem.Name.grantType.rawValue
+            }
+            
+            if containsGrantType {
+                requestBody = body
+            }
+            else {
+                requestBody.append(contentsOf: body)
+            }
+        }
+        
+        return requestBody
+    }
+    
+    private func fetchToken(with request: URLRequest, completionHandler: @escaping (AuthenticationResult) -> Void) {
+        if let url = request.url {
+            self.logger?.log("\(String(describing: self)) will fetch token from url: \(url.absoluteString)", logLevel: .verbose)
+        }
+        
+        URLSession.shared.dataTask(with: request) { data, response, error in
+            if let error = error {
+                self.logger?.log("\(String(describing: self)) failed to retrieve token with error: \(error.localizedDescription)", logLevel: .error)
+                completionHandler(.failure(error: error))
+            }
+            else if let data = data, let token = try? JSONDecoder().decode(OAuthToken.self, from: data) {
```

Comment:
This extension is available only in unit tests target at the moment.

I don't think we want to expose our internal errors to the client. In case of en error we return `AuthenticationError.failedToParseToken` to the client.

Reply:
Valid point about the `default` availability.  `JSONDecoder()` can stay 👍 

Even if we don't pass the underlying parse issue to the client I still think we should log it.  Could prove useful for debugging.